### PR TITLE
Disallow multiple in-progress credential requests

### PR DESCRIPTION
**Pull request by npm1**
Fixes https://github.com/w3c/webappsec-credential-management/issues/206

***
<a href="https://pr-preview.s3.amazonaws.com/npm1/webappsec-credential-management/pull/207.html" title="Last updated on Sep 16, 2022, 4:56 PM UTC (75a6b6a)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webappsec-credential-management/207/8350696...npm1:75a6b6a.html" title="Last updated on Sep 16, 2022, 4:56 PM UTC (75a6b6a)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -942,6 +944,17 @@ spec:css-syntax-3;
 
     1.  Let |p| be [=a new promise=].
 
+    1. Let |topLevelBrowsingContext| be |settings|' [=top-level browsing context=].
+
+    1. [=map/For each=] |optionKey| → <var ignore>optionValue</var> of |options|:
+
+        1. If |topLevelBrowsingContext|'s [=active credential types=] [=set/contains=] |optionKey|,
+            [=reject=] |p| with null and return |p|.
```
Comment (npm1) [inline index.bs:952]: Wasn't sure what we should do (probably not reject p with null?)

Comment (johannhof) [reply index.bs]: Probably an error, see https://www.w3.org/2001/tag/doc/promises-guide#reasons-should-be-errors

`TypeError` seems to be the most generic, or else see https://webidl.spec.whatwg.org/#dfn-error-names-table

File: index.bs [hunk 2]
```diff
@@ -942,7 +944,16 @@ spec:css-syntax-3;
 
     1.  Let |p| be [=a new promise=].
 
-    1.  Let |origin| be the [=current settings object=]'s [=environment settings object/origin=].
+    1. [=map/For each=] |optionKey| → <var ignore>optionValue</var> of |options|:
```
Comment (nsatragno) [inline index.bs:947]: `|optionKey|` here can be `mediation` or `signal`, which are not credential types.

Perhaps

"For each interface in options’ relevant credential interface objects..."

Like [here](https://pr-preview.s3.amazonaws.com/w3c/webappsec-credential-management/207/8350696...npm1:f138311.html#algorithm-collect-known)

File: index.bs [hunk 3]
```diff
@@ -1019,6 +1030,10 @@ spec:css-syntax-3;
             {{CredentialMediationRequirement/conditional}} and a `null` credential is discovered,
             promise |p| is not resolved.
 
+    1. [=map/For each=] |optionKey| → <var ignore>optionValue</var> of |options|:
+
+        1. [=set/Remove=] |optionKey| from |settings|' [=active credential types=].
```
Comment (nsatragno) [inline index.bs:1035]: Clearing the `active credential types` set should happen after the promise is resolved or rejected, not before the promise is returned. The promise is returned immediately, so from the perspective of the website `active credential types` would always be empty, even with a pending request (and therefore, a pending promise).

I think what you want is to [react](https://webidl.spec.whatwg.org/#dfn-perform-steps-once-promise-is-settled) to the promise instead.

File: index.bs [hunk 4]
```diff
@@ -1076,14 +1097,17 @@ spec:css-syntax-3;
...
-            2.  [=Queue a task=] on <var ignore>global</var>'s [=DOM manipulation task source=] to
+            1.  [=Queue a task=] on <var ignore>global</var>'s [=DOM manipulation task source=] to
                 run the following substeps:
 
                 1. [=Reject=] |p| with |e|.
 
             Otherwise, [=resolve=] |p| with |r|.
 
-    6.  Return |p|.
+    1. [=set/Remove=] |credential|'s {{Credential/[[type]]}} from |settings|'
```
Comment (nsatragno) [inline index.bs:1107]: ditto

File: index.bs [hunk 5]
```diff
@@ -1135,22 +1166,24 @@ spec:css-syntax-3;
...
-        3.  Assert: |r| is an algorithm (as defined in [[#algorithm-create-cred]]).
+        1.  Assert: |r| is an algorithm (as defined in [[#algorithm-create-cred]]).
 
-        4.  [=Queue a task=] on |global|'s [=DOM manipulation task source=] to run the following substeps:
+        1.  [=Queue a task=] on |global|'s [=DOM manipulation task source=] to run the following substeps:
 
             1. [=Resolve=] |p| with the result of [=promise-calling=] |r| given |global|.
 
-    11.  Return |p|.
+    1. [=set/Remove=] |type| from |settings|' [=active credential types=].
```
Comment (nsatragno) [inline index.bs:1184]: ditto

**Review by nsatragno — CHANGES_REQUESTED** [review]

File: index.bs (hunk 2 above)
Comment (npm1) [reply index.bs]: Done

File: index.bs (hunk 3 above)
Comment (npm1) [reply index.bs]: Done

File: index.bs (hunk 4 above)
Comment (npm1) [reply index.bs]: Done

File: index.bs (hunk 5 above)
Comment (npm1) [reply index.bs]: Done

**Review by npm1 — COMMENTED** [review]
I forgot to disable autoremoval of blank spaces before end of a line...

**Comment by nsatragno** [comment]
It looks like this PR needs to be rebased, there are a bunch of unrelated changes (like the introduction of permissions policy for fedcm).

**Comment by npm1** [comment]
> It looks like this PR needs to be rebased, there are a bunch of unrelated changes (like the introduction of permissions policy for fedcm).

What do you mean? https://github.com/npm1/webappsec-credential-management/tree/repeated says I'm already ahead of w3c:main branch. I skimmed the diff and didn't see the introduction of permissions policy.

File: index.bs [hunk 6]
```diff
@@ -942,12 +942,29 @@ spec:css-syntax-3;
 
     1.  Let |p| be [=a new promise=].
 
-    1.  Let |origin| be the [=current settings object=]'s [=environment settings object/origin=].
+    1. For each |interface| in |options|' <a>relevant credential interface objects</a>:
+
+        1. If |settings|' [=active credential types=] [=set/contains=] |interface|'s
+            {{Credential/[[type]]}}, return [=a promise rejected with=] a "{{NotAllowedError}}"
```
Comment (nsatragno) [inline index.bs:948]: Unrelated to the PR but can you file a crbug for this & to have a corresponding WPT?

**Review by nsatragno — APPROVED** [review]
LGTM, thanks

File: index.bs (hunk 6 above)
Comment (npm1) [reply index.bs:948]: Filed https://bugs.chromium.org/p/chromium/issues/detail?id=1365979

**Pull request merged**